File: java语言/src/DataStructure/LinearStructure/LinkedList/LinkList.java
package DataStructure.LinearStructure.LinkedList;

/*
* 带头节点单链表的实现
* */
class Data{
    String username;
    int age;

    public Data(String username, int age) {
        this.username = username;
        this.age = age;
    }

    @Override
    public String toString() {
        return "Data{" +
                "username='" + username + '\'' +
                ", age=" + age +
                '}';
    }
}
class Node{

    Data data;
    private Node next;

    public Node(Data data, Node next) {
        this.data = data;
        this.next = next;
    }

    public Node() {
    }

    public Node getNext() {
        return next;
    }

    public void setNext(Node next) {
        this.next = next;
    }
}
public class LinkList {
    private Node node;
    private Node cruNode;
    //初始化一个链表
    public void init(){
        Node node=new Node();
         this.node=node;
         this.cruNode=node;
    }
    //尾结点插入
    public LinkList putNodeByTail(String username,int age){
        Data data = new Data(username,age);
        Node newNode=new Node(data,null);
        cruNode.setNext(newNode);
        cruNode=newNode;
        return this;
    }
    //头节点插入
    public void putNodeByHead(String username,int age){
        Data data = new Data(username,age);
        Node newNode=new Node(data,null);
        if (node!=null){
            Node tmp=node.getNext();
            newNode.setNext(tmp);
            node.setNext(newNode);
        }
    }
    //删除其中一个结点
    public void deleteNode(int i){
        if (i==1){
            Node next = node.getNext().getNext();
            node.setNext(next);
        }else {
            int count=2;
            Node froNode=node.getNext();
            Node preNode=node.getNext().getNext();
            while (preNode!=null){
                if (count==i){
                    froNode.setNext(preNode.getNext());
                }
                froNode=preNode;
                preNode=preNode.getNext();
                ++count;
            }
        }
    }
    //向链表中插入元素
    public void insertNode(int i,String username,int age){
        Data data = new Data(username, age);
        Node newNode = new Node(data, null);
        int count=1;
        Node preNode=node.getNext();
        Node froNode=null;
        while (preNode!=null){
            if (count==i){
                froNode.setNext(newNode);
                newNode.setNext(preNode);
            }
            froNode=preNode;
            preNode=preNode.getNext();
            ++count;
        }
    }
    //判断链表是否为空
    public boolean isEmpty(){
        return node.getNext()!=null?false:true;
    }
    //获取第i个结点
    public Node getNode(int i){
        Node countNode=node.getNext();
        int count=1;
        while (countNode!=null){
            if (count==i){
                return countNode;
            }
            countNode=countNode.getNext();
            ++count;
        }
        if (countNode==null) return null;
        return countNode;
    }

    //打印链表中的数据
    public void printNodeData(){
        Node preNode=node.getNext();
        while (preNode!=null){
            System.out.println(preNode.data.toString());
            preNode=preNode.getNext();
        }
    }
}

File: C语言/LinkList.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>	

typedef struct Book {
	char bookname[20];
	char author[10];
}Book;

typedef struct Node {
	Book book;
	struct Node *next;
}Node,*LinkList;

//初始化链表 
void initChainedTable(Node **L){    //ok
	*L=(LinkList)malloc(sizeof(Node));
	(*L)->next=NULL;
}
//尾结点插入 
Node* putNodeByTail(Book book,Node *r){   //ok
	Node *node=(Node*)malloc(sizeof(Node));
	node->book= book;
	node->next=NULL;
	r->next=node;
	return node;
}
//首节点插入
void putNodeByHead(Book book,LinkList L){   //ok
	Node *node=(Node*)malloc(sizeof(Node));
	node->book=book;
	node->next=L->next;
	L->next=node;
} 
//删除其中一个结点
void deleteNode(int i,LinkList L){   //ok
	Node *node=L->next;
	int counter=1;
	if (i == 1) {
        if (node != NULL) {
            L->next = node->next;
            free(node);
        } else {
            printf("List is empty, cannot delete.\n");
        }
        return;
    }

	while(node&&counter<(i-1)){
		node=node->next;
		++counter;
	}
	if(!node||i<=0){
		printf("index is out of bound \n");
	}else{
		Node* delnode = node->next;
        if (delnode) {
            node->next = delnode->next;
            free(delnode);
        } else {
            printf("Node at index %d does not exist\n", i);
        }
	}
	
}
//向链表中插入元素
void insertNode(Book book,int i,LinkList L){   //ok
	Node *newnode=(Node*)malloc(sizeof(Node));
	Node *node=L->next;
	newnode->book=book;
	int count=1;
	while(node&&count<(i-1)){
		node=node->next;
		++count;
	}
	if(!node||i<=0){
		printf("index is out of bound\n");
	}else{
		newnode->next=node->next;
		node->next=newnode;	
	}
	
} 
//判断链表是否为空
int isEmpty(LinkList L){  //ok
	return L->next==NULL?1:0;
} 

//获取第i个结点 
Node*  getNode(int i,LinkList L){  //ok
	Node* node=L->next;
	int count=1;
	while(node&&count<i){
		node=node->next;
		++count;
	}
	if(node&&i>0) return node;
	printf("index is out of bound\n"); 
} 
//通过值获取到它的位置
int getIndex(Book book,LinkList L){  //ok
	Node* node=L->next;
	int count=1;
	while(node&&strcmp(node->book.author,book.author)&&strcmp(node->book.bookname,book.bookname)){
		node=node->next;
		++count;
	}
	if(node) return -1;
	return count;
} 
//打印链表中的数据 
void printNodeData(LinkList L){    //ok
	Node *node=L->next;
	while(node){
		printf("bookname:%s;author:%s\n",node->book.bookname,node->book.author);
		node=node->next;
	}
	if(!node) printf("你所给的是个空链表！\n"); 
}
//释放一个链表
void freeChainedTable(LinkList L){ 
	Node *p=L->next;
	Node *q;
	while(p){  
		q=p->next;
		free(p);
		p=q; 
	}
	printf("清除完成！\n");
	L->next=NULL;
}


int main(){
	LinkList L;
	initChainedTable(&L);
	Book book1={"java","zhangyu"};
	Book book2={"php","zhang"};
	Book book3={"c#","nanchengyu"};
	putNodeByHead(book1,L);
	putNodeByHead(book2,L);
	putNodeByHead(book3,L);
	Book book4={"java","zhang"};
	freeChainedTable(L);
	printNodeData(L);
	system("pause");
	return 0;
}
